add simple-build-dgcode package

tkittel opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| Rust            | `@conda-forge/help-rust`      |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Gitter channel][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io

All apologies in advance if your recipe PR does not receive prompt attention.
This is a high volume repository and the reviewers are volunteers. Review times
vary depending on the number of reviewers on a given language team and may be days
or weeks. We are always looking for more staged-recipe reviewers. If you are
interested in volunteering, please contact a member of @conda-forge/core.
We'd love to have the help!
-->

Checklist
- [ ] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [ ] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [ ] Source is from official source.
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [ ] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [ ] Build number is 0.
- [ ] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

tkittel commented:
@conda-forge/help-python ready for review!

tkittel commented:
Everything rebased and version req. updated in meta.yaml.


@conda-forge/help-python ready for review!

dhirschfeld commented on recipes/simple-build-dgcode/build.sh:
```diff
@@ -0,0 +1,13 @@
+#To follow conda-forge's best practices, we got the sources via a tarball, but
+#simple-build-dgcode uses setuptools-git-versioning which gets the version from
+#the git tag. So we initialise a git repo and add the version tag.
+git init
```
If the project uses a versioning tool which doesn't work with sdists I'd be inclined to just use a git checkout rather than this hack.

@conda-forge/help-python may not agree with that? I'm not sure how religiously that "requirement" is supposed to be taken.

xhochy replied:
This tool probably generates an `sdist` where the tag is baked in. You should be using the sdist instead of the git tarball as the source to build from.

jaimergp replied:
PyPI is the canonical source for Python packages, whenever possible. PyPI being down is as unlikely as catastrophical; we'd have bigger problems if that happens :)

Yes, the bots understand PyPI URL schemes. Follow the example provided for context.

tkittel replied:
Ok, I have removed the "git init" hack and now use the pypi sdist package as the source :-)

dhirschfeld reviewed (approved):
Other than the versioning hack, this looks ok to me.

As for the hack, I don't think it should preclude this being merged if that's something you want to support in your feedstock, but hopefully it's not necessary.

tkittel commented:
@xhochy and @jaimergp based on your input I have now rewritten the recipe to simply use a pypi sdist archive as the source. I also took the opportunity and rebased the PR on the latest upstream.

So I hope this resolves everything? :-)

tkittel commented:
@xhochy and @jaimergp are you happy now to merge this? Basically your recommendation was spot-on, and I simply adapted it and everything worked smoothly in one go :-)

jaimergp commented on recipes/simple-build-dgcode/run_test.sh:
```diff
@@ -0,0 +1,14 @@
+dgcode_newsimproject TriCorder
+sb --tests --testexcerpts=10000
+sbenv sb_core_extdeps --require NCrystal Geant4
+
+
+
```
```suggestion

```

jaimergp commented on recipes/simple-build-dgcode/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+    - dgcode_newsimproject = simplebuild_dgcode._cli:newsimproject
+
+requirements:
+  build:
+    - python                                 # [build_platform != target_platform]
+    - cross-python_{{ target_platform }}     # [build_platform != target_platform]
```
```suggestion
```

I don't think we need to cross-compile anything here 🤔

tkittel replied:
Right, I will remove them.

jaimergp commented on recipes/simple-build-dgcode/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+    - python                                 # [build_platform != target_platform]
+    - cross-python_{{ target_platform }}     # [build_platform != target_platform]
+  host:
+    - python
```
```suggestion
    - python >=3.8
```

tkittel replied:
Ok, adding that to both host and run requirements then (I don't think this package will see any usage below python 3.9 or even 3.10, so in practice it probably won't matter).

jaimergp commented on recipes/simple-build-dgcode/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+    - setuptools-git-versioning
+    - setuptools
+  run:
+    - python
```
```suggestion
    - python >=3.8
```

tkittel replied:
Will do.

tkittel commented on recipes/simple-build-dgcode/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+{% set version = "0.0.12" %}
+
+#NB get sha256 with: curl -sL <URL> | openssl sha256
```
Ok, thanks, I will update the comment!

tkittel commented on recipes/simple-build-dgcode/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+  sha256: fdde85d7b07dda6b11e6b28bcdd2510c45afdd2d976a096640cf86f63adc0a75
+
+build:
+  skip: True  # [win or py<39]
```
@jaimergp I will try, but I am not sure it will work since the simple-build-system package (a dependency) is not available on windows.

tkittel replied:
@jaimergp Ok, great, that would be wonderful. I think I abandoned previous efforts since the window build failed and then I did not get all tests passed. But do you not think that this will cause problems in the future when I want to have to bots automatic update the versions (i.e. in those automatic bot PRs, will there not also be a spurious test failure for the windows builds?).

tkittel replied:
(I added the same question in the main conversation thread of this PR, with a few more details)

tkittel commented:
@jaimergp I now implemented your suggested changes, and as expected the windows test fails while osx and linux passes.

Does this mean everything is good now? I am a bit curious, because in the azure logs, I see the final package hashes as:

```
{
2024-02-23T11:03:28.7215384Z   "simple-build-dgcode-0.0.15-pyh4af843d_0.conda": {
2024-02-23T11:03:28.7215618Z     "recipe": {
2024-02-23T11:03:28.7215804Z       "c_compiler": "gcc",
2024-02-23T11:03:28.7216022Z       "c_compiler_version": "12",
2024-02-23T11:03:28.7216301Z       "channel_targets": "conda-forge main",
2024-02-23T11:03:28.7216526Z       "cxx_compiler": "gxx",
2024-02-23T11:03:28.7217056Z       "cxx_compiler_version": "12"
2024-02-23T11:03:28.7217218Z     }
2024-02-23T11:03:28.7217336Z   }
2024-02-23T11:03:28.7217475Z }
```
for linux, and on osx:
```
2024-02-23T11:13:52.3946210Z {
2024-02-23T11:13:52.3947460Z   "simple-build-dgcode-0.0.15-pyh2f45acb_0.conda": {
2024-02-23T11:13:52.3954830Z     "recipe": {
2024-02-23T11:13:52.3955920Z       "CONDA_BUILD_SYSROOT": "/Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk",
2024-02-23T11:13:52.3956990Z       "c_compiler": "clang",
2024-02-23T11:13:52.3957900Z       "c_compiler_version": "16",
2024-02-23T11:13:52.3959020Z       "channel_targets": "conda-forge main",
2024-02-23T11:13:52.3959990Z       "cxx_compiler": "clangxx",
2024-02-23T11:13:52.3960930Z       "cxx_compiler_version": "16"
2024-02-23T11:13:52.3961840Z     }
2024-02-23T11:13:52.3962700Z   }
2024-02-23T11:13:52.3963580Z }
```

Indeed, there is no mention of python version, so I guess that is good. There is still some compiler stuff, which I am not sure about how gets there. But if you think this is good/normal, then I am happy. Otherwise I would love to hear ideas for what to do about it :-)

And as I mentioned in another comment, is the failed windows build here simply a feature of the staged-recipes repo, and something that would be absent in the generated feedstock repo? Because, otherwise won't the automatic PRs generated by the bots upon new releases also have windows build failures, and therefore won't get merged automatically?

jaimergp commented on recipes/simple-build-dgcode/meta.yaml:
```diff
@@ -0,0 +1,55 @@
+test:
+  requires:
+    # Note: The actual tests are defined in run_test.sh
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
```
```suggestion
    - c-compiler
    - cxx-compiler
```

Let's try these to get rid of those hashes.

tkittel replied:
Thanks! I had absolutely no idea that packages in the test requirements would affect the hash of the package :-)

jaimergp commented:
In feedstocks, noarch packages only get built on Linux (that's one of their advantages), so you won't see errors on Windows because it won't run.

tkittel commented:
@jaimergp that certainly helped to remove the version numbers from the hashes:
```
2024-02-23T14:30:04.1001688Z {
2024-02-23T14:30:04.1002161Z   "simple-build-dgcode-0.0.15-pyh6459364_0.conda": {
2024-02-23T14:30:04.1002367Z     "recipe": {
2024-02-23T14:30:04.1002570Z       "c_compiler": "gcc",
2024-02-23T14:30:04.1002834Z       "channel_targets": "conda-forge main",
2024-02-23T14:30:04.1003056Z       "cxx_compiler": "gxx"
2024-02-23T14:30:04.1003238Z     }
2024-02-23T14:30:04.1003415Z   }
2024-02-23T14:30:04.1003595Z }
```
and for osx:
```
2024-02-23T14:37:19.9181110Z {
2024-02-23T14:37:19.9182290Z   "simple-build-dgcode-0.0.15-pyh82e1e4f_0.conda": {
2024-02-23T14:37:19.9183010Z     "recipe": {
2024-02-23T14:37:19.9183650Z       "CONDA_BUILD_SYSROOT": "/Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk",
2024-02-23T14:37:19.9184280Z       "c_compiler": "clang",
2024-02-23T14:37:19.9184910Z       "channel_targets": "conda-forge main",
2024-02-23T14:37:19.9185510Z       "cxx_compiler": "clangxx"
2024-02-23T14:37:19.9185980Z     }
2024-02-23T14:37:19.9186420Z   }
2024-02-23T14:37:19.9186790Z }
```
So it is a lot nicer without the compiler's version number, but there is still a difference between linux and osx hashes. Do you think this might be caused by a dependency (e.g. the simple-build-system package, which also has a `{{ compiler('c') }}` in the test requirement section).?

jaimergp commented:
No clue but I don't think it's important. We will only generate the linux variant anyway, and that will be used in all OSs.

tkittel commented:
@jaimergp Ok, thanks a lot for the spot-on advice really very much appreciated! :-)
